src/datasets.c

```diff
@@ -156,6 +163,217 @@ static int ParseRepLine(const char *in, size_t ins, DataRepType *rep_out)
+{
+    if (strlen(set->load) == 0)
+        return 0;
+
+    SCLogConfig("dataset: %s loading from '%s'", set->name, set->load);
+    const char *fopen_mode = "r";
+    if (strlen(set->save) > 0 && strcmp(set->save, set->load) == 0) {
+        fopen_mode = "a+";
+    }
+
+    FILE *fp = fopen(set->load, fopen_mode);
+    if (fp == NULL) {
+        SCLogError(SC_ERR_DATASET, "fopen '%s' failed: %s", set->load, strerror(errno));
+        return -1;
+    }
+
+    uint32_t cnt = 0;
+    char line[1024];
+    while (fgets(line, (int)sizeof(line), fp) != NULL) {
+        char *r = strchr(line, ',');
+        if (r == NULL) {
+            line[strlen(line) - 1] = '\0';
+            SCLogDebug("line: '%s'", line);
+
+            struct in_addr in;
+            if (inet_pton(AF_INET, line, &in) != 1) {
+                FatalErrorOnInit(SC_ERR_FATAL, "dataset data parse failed %s/%s: %s", set->name,
+                        set->load, line);
+                continue;
+            }
+
+            if (DatasetAdd(set, (const uint8_t *)&in.s_addr, 4) < 0) {
+                FatalErrorOnInit(
+                        SC_ERR_FATAL, "dataset data add failed %s/%s", set->name, set->load);
+                continue;
+            }
+            cnt++;
+
+            /* list with rep data */
+        } else {
+            line[strlen(line) - 1] = '\0';
+            SCLogDebug("IPv4 with REP line: '%s'", line);
+
+            *r = '\0';
+
+            struct in_addr in;
+            if (inet_pton(AF_INET, line, &in) != 1) {
+                FatalErrorOnInit(SC_ERR_FATAL, "dataset data parse failed %s/%s: %s", set->name,
+                        set->load, line);
+                continue;
+            }
+
+            r++;
+
+            DataRepType rep = { .value = 0 };
+            if (ParseRepLine(r, strlen(r), &rep) < 0) {
+                FatalErrorOnInit(SC_ERR_FATAL, "bad rep for dataset %s/%s", set->name, set->load);
+                continue;
+            }
+
+            SCLogDebug("rep v:%u", rep.value);
+            if (DatasetAddwRep(set, (const uint8_t *)&in.s_addr, 4, &rep) < 0) {
+                FatalErrorOnInit(
+                        SC_ERR_FATAL, "dataset data add failed %s/%s", set->name, set->load);
+                continue;
+            }
+
+            cnt++;
+        }
+    }
+    THashConsolidateMemcap(set->hash);
+
+    fclose(fp);
+    SCLogConfig("dataset: %s loaded %u records", set->name, cnt);
+    return 0;
+}
+
+static int ParseIpv6String(Dataset *set, char *line, struct in6_addr *in6)
+{
+    /* Checking IPv6 case */
+    char *got_colon = strchr(line, ':');
+    if (got_colon) {
+        if (inet_pton(AF_INET6, line, in6) != 1) {
+            FatalErrorOnInit(SC_ERR_FATAL, "dataset data parse failed %s/%s: %s", set->name,
+                    set->load, line);
+            return -1;
+        }
+        /* IPv4 in IPv6 notation needs transformation to internal Suricata storage */
+#ifndef OS_WIN32
```

Comment:
can we hide this kind of ifdeffery?

Reply:
we do address parsing elsewhere w/o needing this?

Reply:
I tried to find and did not find it.

Reply:
did you look at `detect-engine-address.c`

Reply:
Indeed it can be used.